[Core] Add fallback strategy scheduling logic

ryanaoleary opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR also updates the cluster resource scheduler logic to account for the list of `LabelSelector`s specified by the `fallback_strategy`, falling back to each fallback strategy `LabelSelector` in-order until one is satisfied when selecting the best node. We're able to support fallback selectors by considering them in the cluster resource scheduler in-order using the existing label selector logic in `IsFeasible` and `IsAvailable`, returning the first valid node returned by `GetBestSchedulableNode`.

## Related issue number

https://github.com/ray-project/ray/issues/51564

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ryanaoleary committed:
Add fallback strategy initial commit

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

Add tests and fix scheduling logic

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

remove cgroup change

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

Fix merge

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented:
The python changes are in this separate PR: https://github.com/ray-project/ray/pull/56374 which can be merged first. This PR will then contain only the C++ scheduling logic changes. cc: @MengjinYan

MengjinYan commented on src/ray/raylet/scheduling/cluster_resource_scheduler.cc:
```diff
@@ -289,58 +289,77 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
+      // If the task is being scheduled by gcs, return nil to make it stay in the
+      // `cluster_lease_manager`'s queue.
+      if (!is_local_node_with_raylet_) {
+        return scheduling::NodeID::Nil();
```
Wondering in this case, should we check the fallback as well? 

cc: @jjyao to see if you have more context on the corner case where the scheduling happens on GCS and the `GetBestSchedulableNode` returns not till while the returned node is not schedulable.

MengjinYan reviewed (commented):
Haven't looked at the tests yet and there are also seems to be conflicts that need to be fixed.

ryanaoleary committed:
Fix lint

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

MengjinYan commented on python/ray/_private/label_utils.py:
```diff
@@ -189,3 +190,17 @@ def validate_node_label_syntax(labels: Dict[str, str]):
+            raise TypeError("Each element in fallback_strategy must be a dictionary.")
+        possible_error_message = validate_label_selector(label_selector)
+        if possible_error_message:
+            return possible_error_message
```
Looks like a valid comment.

ryanaoleary replied:
Fixed this in the PR with just the API change, I'll remove these changes from this PR once the other one is merged and I'm able to rebase on it. I've been getting weird pydoc CI errors that don't show when I run the scripts locally.

ryanaoleary committed:
Update python/ray/tests/test_label_scheduling.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Fix merge

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

MengjinYan commented on python/ray/_raylet.pyx:
```diff
@@ -855,6 +855,32 @@ cdef int prepare_label_selector(
+
+        c_label_selector_map.clear()
+        prepare_label_selector(label_selector, &c_label_selector_map)
+        fallback_strategy_vector.push_back(c_label_selector_map)
```
Just curious, in the future when we add the resource fallback, how do we plan to update the logic to add the additional resource fallbacks?

ryanaoleary replied:
In 8c71778 I added a `FallbackStrategy` struct we can easily add to by simply adding more fields. It should be as simple as adding resource fallback as a supported option, adding the new field in the C++ struct, and ensuring that it's parsed and then considered wherever fallback strategy is used.

ryanaoleary committed:
Update python/ray/remote_function.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_common/ray_option_utils.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary commented on python/ray/_raylet.pyx:
```diff
@@ -855,6 +855,32 @@ cdef int prepare_label_selector(
+        c_label_selector_map.clear()
+        prepare_label_selector(label_selector, &c_label_selector_map)
+        fallback_strategy_vector.push_back(c_label_selector_map)
+
```
This is intended behavior, we expect to pass an empty label selector {} if None is specified.

ryanaoleary committed:
Refactor to correct fallback_strategy format, and parse to C++ type earlier

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

ryanaoleary committed:
Fix cluster resource scheduler test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

MengjinYan commented on src/ray/common/task/task_util.h:
```diff
@@ -187,8 +191,11 @@ class TaskSpecBuilder {
+    *message_->mutable_label_selector() = label_selector.ToProto();
+    for (const auto &strategy_map : fallback_strategy) {
+      *message_->add_fallback_strategy()->mutable_label_selector() =
+          strategy_map.label_selector.ToProto();
```
Probably we should add `ToProto` function also to the fallback strategy options, so that when we add new fields in the construct we can just modify the struct/class itself.

MengjinYan commented on src/ray/protobuf/public/events_actor_definition_event.proto:
```diff
@@ -39,5 +41,5 @@ message ActorDefinitionEvent {
   bytes placement_group_id = 9;
   // The label selector for the actor.
-  map<string, string> label_selector = 11;
+  LabelSelector label_selector = 11;
```
This is a public proto schema. So we probably shouldn't modify it but keep it as `map<string, string>`

ryanaoleary replied:
Added in f873181

ryanaoleary commented:
cc: @alanwguo for review since I changed some of the protos related to https://github.com/ray-project/ray/pull/53423 for the label selector field, since we now parse it from a map sooner in the code.

ryanaoleary committed:
Fix types in related tests

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

MengjinYan commented on src/ray/protobuf/common.proto:
```diff
@@ -597,9 +598,11 @@ message TaskSpec {
+  LabelSelector label_selector = 43;
   // The tensor transport type for this task.
   TensorTransport tensor_transport = 44;
+  // A list of label selectors defining the fallback strategy for scheduling.
```
```suggestion
  // A list of fallback options defining the fallback strategy for scheduling.
```

MengjinYan commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -1997,7 +1999,9 @@ std::vector<rpc::ObjectReference> CoreWorker::SubmitTask(
-                      task_options.label_selector);
+                      task_options.label_selector,
+                      task_options.fallback_strategy,
+                      task_options.tensor_transport);
```
Nit: Wondering why do we need the `task_options.tensor_transport` here?

MengjinYan reviewed (commented):
Thanks for the refactor!! 

Added some comments for the core logic. Still need to look at whether the change of the actor table and task into entry will impact observability feature and the test.

MengjinYan reviewed (commented):
Followup comments on the implication of  proto schema changes.

ryanaoleary committed:
Update src/ray/protobuf/common.proto

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update src/ray/common/scheduling/fallback_strategy.h

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/remote_function.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
parse label selector to map for public protos

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Change to FallbackOption and fix other comments

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
add fallback strategy test and fix names

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on src/ray/protobuf/common.proto:
```diff
@@ -597,9 +598,11 @@ message TaskSpec {
   // The tensor transport type for this task.
   TensorTransport tensor_transport = 44;
+  // A list of fallback options defining the fallback strategy for scheduling.
+  repeated FallbackStrategyOptions fallback_strategy = 45;
```
Makes sense to me, done in [f873181](https://github.com/ray-project/ray/pull/56369/commits/f873181ecd0af061062fafbfea55a6874a08658a)

ryanaoleary commented on src/ray/protobuf/gcs.proto:
```diff
@@ -159,7 +159,7 @@ message ActorTableData {
   optional string call_site = 34;
   // The label selector of the actor.
-  map<string, string> label_selector = 35;
+  LabelSelector label_selector = 35;
```
Done in [f873181](https://github.com/ray-project/ray/pull/56369/commits/f873181ecd0af061062fafbfea55a6874a08658a), it now remains a string map

ryanaoleary commented on python/ray/includes/common.pxd:
```diff
@@ -346,8 +357,9 @@ cdef extern from "ray/core_worker/common.h" nogil:
-                     CTensorTransport tensor_transport)
+                     const CLabelSelector &label_selector,
+                     CTensorTransport tensor_transport,
+                     const c_vector[CFallbackStrategyOptions] &fallback_strategy)
```
I don't think there was a reason for it, I changed it to match the c++ constructor in [fbee074](https://github.com/ray-project/ray/pull/56369/commits/fbee074208a6fa63861da802bfa89c4d8b15231f)

ryanaoleary commented:
> So I'm thinking, if we consider the fallback integration in general, we should do the following:

> When the node is feasible but not available on the best node, I think we should go and check the fallback options to see if there is anything that can be scheduled right away
At the same time, there could still be a case where, all the fallback options are unavailable, then in this case, probably we should go with the best node that satisfies the highest priority fallback options
Then in the case of scheduling on GCS, if all fallback options are unavailable, we should return nil.
What do you think?


Yeah that makes sense to me, I refactored to cover these cases in[ff8fcc6](https://github.com/ray-project/ray/pull/56369/commits/ff8fcc6f1256d1e210122f1679bfc64e8e0be4f6) and added tests for each possible path.

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

ryanaoleary committed:
run format

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix bugbot comments

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on src/ray/raylet/scheduling/cluster_resource_scheduler.cc:
```diff
@@ -289,58 +289,98 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
+
+    if (cluster_resource_manager_->HasFeasibleResources(local_node_id_,
+                                                        resource_request)) {
+      return local_node_id_;
```
[ca29d79](https://github.com/ray-project/ray/pull/56369/commits/ca29d79506a6c72c808e46a3b083bd2635b6eb20)

ryanaoleary committed:
fix var names

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

ryanaoleary committed:
ensure to string map supports ! operator

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

MengjinYan commented on src/ray/raylet/scheduling/cluster_resource_scheduler.cc:
```diff
@@ -289,58 +289,107 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
+    const auto &selector_for_local_check =
+        highest_priority_unavailable_label_selector.has_value()
+            ? highest_priority_unavailable_label_selector.value().get()
+            : lease_spec.GetLabelSelector();
```
Curious if there is a case that `highest_priority_unavailable_label_selector.has_value()` is false? If not, I think we can add a `RAY_CHECK` here if we want to catch all cases

MengjinYan reviewed (commented):
Looks good overall on the main logic! Just minor comments. 

Thanks for updating the scheduling logic!! It is much cleaner now. 

Reviewing the tests now.

ryanaoleary committed:
Update src/ray/common/scheduling/fallback_strategy.cc

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update src/ray/protobuf/public/events_actor_definition_event.proto

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update src/ray/raylet/scheduling/cluster_resource_scheduler.cc

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

MengjinYan reviewed (commented):
Just 1 minor comments on the tests

ryanaoleary committed:
Fix label selector ref

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Add more tests cases, fix SerializeFallbackStrategy, and rename to FallbackOption

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on src/ray/common/scheduling/fallback_strategy.cc:
```diff
@@ -0,0 +1,52 @@
+  return strategy_list;
+}
+
+std::unique_ptr<rpc::FallbackStrategy> SerializeFallbackStrategy(
```
Done in fae8440, I changed it to return a `rpc::FallbackStrategy`

ryanaoleary committed:
add back sorting for deterministic tostringmap output

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Use sorted values

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

MengjinYan reviewed (approved):
Thanks! Just one minor followup question on the ToProto function.

ryanaoleary committed:
Change label selector proto to directly write to object, and add test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

ryanaoleary committed:
Fix e2e test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix test incorrectly creating infeasible nodes

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Merge branch 'master' into fallback-selector-api

MengjinYan commented:
cc: @jjyao @edoakes

edoakes commented on python/ray/tests/test_label_scheduling.py:
```diff
@@ -96,5 +96,92 @@ def test_label_selector_multiple(cluster_with_labeled_nodes):
+        node_1,
+        node_2,
+        node_3,
+    }
```
consider making the test more deterministic by scheduling 3 actors in parallel, each that occupies all CPUs on each node, and asserting that all 3 nodes are occupied by one of the actors

edoakes commented on src/ray/raylet/scheduling/cluster_resource_scheduler.cc:
```diff
@@ -289,58 +289,103 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
+  // Construct list of references to all LabelSelectors, from both the `label_selector`
+  // and `fallback_strategy` arguments.
+  std::vector<std::reference_wrapper<const LabelSelector>> label_selectors;
+  label_selectors.push_back(std::cref(lease_spec.GetLabelSelector()));
```
I've never seen reference_wrapper and `std::cref` before -- is this just a special way to have a vector of const refs and avoid copying into the vector?

ryanaoleary replied:
Yeah exactly, `std::vector<const LabelSelector&>` wouldn't compile so I found those helpers. I only wanted to store the references to avoid copying unnecessarily into 1 list when we already had the original objects.

edoakes commented on src/ray/raylet/scheduling/cluster_resource_scheduler.cc:
```diff
@@ -289,58 +289,103 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
+        requires_object_store_memory);
+
+    // Use the label selector from the highest-priority fallback that was feasible.
+    // There must be at least one feasible node and selector.
```
```suggestion
    // There must be at least one feasible node and selector, else we would have returned early above.
```

edoakes reviewed (approved):
Looks great. Only nits that can be addressed in follow up PRs.
